packages/amplify_datastore_plugin_interface/lib/src/types/models/query_snapshot.dart

```diff
@@ -89,44 +110,60 @@ class QuerySnapshot<T extends Model> {
   QuerySnapshot<T> withSubscriptionEvent({
     required SubscriptionEvent<T> event,
   }) {
-    SortedList<T> sortedListCopy = SortedList.from(this._sortedList);
-    SortedList<T>? updatedSortedList;
-
-    T newItem = event.item;
-    bool newItemMatchesPredicate = where == null || where!.evaluate(newItem);
-    int currentItemIndex =
-        sortedListCopy.indexWhere((item) => item.getId() == newItem.getId());
-    T? currentItem =
-        currentItemIndex == -1 ? null : sortedListCopy[currentItemIndex];
-    bool currentItemMatchesPredicate =
-        currentItem != null && (where == null || where!.evaluate(currentItem));
-
-    if (event.eventType == EventType.create &&
-        newItemMatchesPredicate &&
-        currentItem == null) {
-      updatedSortedList = sortedListCopy..addSorted(newItem);
-    } else if (event.eventType == EventType.delete && currentItem != null) {
-      updatedSortedList = sortedListCopy..removeAt(currentItemIndex);
-    } else if (event.eventType == EventType.update) {
-      if (currentItemMatchesPredicate &&
-          newItemMatchesPredicate &&
-          currentItem != newItem) {
-        updatedSortedList = sortedListCopy
-          ..updateAtSorted(currentItemIndex, newItem);
-      } else if (currentItemMatchesPredicate && !newItemMatchesPredicate) {
-        updatedSortedList = sortedListCopy..removeAt(currentItemIndex);
-      } else if (currentItem == null && newItemMatchesPredicate) {
-        updatedSortedList = sortedListCopy..addSorted(newItem);
-      }
-    }
-    if (updatedSortedList != null) {
-      return QuerySnapshot._(
-        sortedList: updatedSortedList,
-        isSynced: isSynced,
-        where: where,
-        sortBy: sortBy,
-      );
+    T item = event.item;
+    String id = item.getId();
+    bool itemMatchesPredicate = where == null || where!.evaluate(item);
+    bool snapshotContainsItem = _models.containsKey(id);
+
+    switch (event.eventType) {
+      case EventType.create:
+        // if the newly created item matches the predicate and the item is
+        // not already in the list, return a new snapshot with the new item
+        if (itemMatchesPredicate && !snapshotContainsItem) {
+          return this.copyWith(
+            sortedList: SortedList.from(_sortedList)..addSorted(item),
+            models: Map.from(_models)..updateItem(id, item),
+          );
+        }
+        break;
+      case EventType.update:
+        // if the updated item is in the list, update or remove based
+        // on the the query predicate
+        if (snapshotContainsItem) {
+          T currentItem = _models[id]!;
+          if (itemMatchesPredicate && currentItem != item) {
```

Comment:
`currentItem != item` may cause ignoring pushing cloud synced model into the snapshot.
See https://github.com/aws-amplify/amplify-flutter/issues/1355#issuecomment-1036669491.